### ISSUE TITLE
What's the right/best way to make fairseq optimizer support different learning rate for different part

### ISSUE THREAD

**Issue by memeda** [pytorch/fairseq#2194]
## ❓ Different Learning Rate for different Model part

for PyTorch, in my current view, there exists two ways to support different learning rate for different part:

1. use different optimizer for different part
2. use [per-parameter options](https://pytorch.org/docs/stable/optim.html#per-parameter-options)

But I did't got any help(search the doc, only api; search issue, no relates) on how to do this under fairseq framework.

In my naive option, I found override the `set_lr` of `FairseqOptimizer` derived class may be the easiest way (that is, we use the `per-parameter options`), but I haven't try and don't know whether it is recommended. And I want to get help on this. 

## update

I found it needs more info: when enable fp16 training, there exists a **Flatten Grads** operator, 
it flatten all parameters to one param(by default), and use it to initialize the Optimizer. And under this condition, the Optimizer only got 1 parameter, so a different learning-rate is impossible.
So we may change  `fp16_no_flatten_grads` to `True` from the default `False`. But how this matters?  in `options.py`, it only has `help='don\'t flatten FP16 grads tensor'`, In my option, it will only affect the speed? 

## update 2

Oh, I think I can't make this without change the fairseq code...

**Comment by myleott** [comment]
Unfortunately this isn’t well supported right now. I can look into adding support for this 

Another option is to modify the magnitude of the gradient using the GradMultiply module. For example we use this to lower the effective learning rate for the convolutional NMT model: https://github.com/pytorch/fairseq/blob/2f7e3f33235b787de2e34123d25f659e34a21558/fairseq/models/fconv.py#L268-L269

**Comment by memeda** [comment]
Thanks!
This method is a bit tricky, and I don't know what's will it affect.
Let's say:
A model is composited by 
```
A -> B -> C
```
I want to let part **B** to get lr1, and let **A**, **C** share the same lr.
following the given method, I can could done by 

```python
def forward(self, xxxx):
     # let's say B part's lr1 == 0.2, and A, C part lr2 = 0.02, so it has 10 times.
     x = A()
     x = GradMultiply.apply(x, 1 / 10) # scale to 1/10 based on B => lr = lr2 = 0.02
     x = B()
     x = GradMultiply.apply(x, 10) # scale to 10 based on C => lr = lr1 = 0.2
     x = C()
```

Is my understanding right?

I'm now try to modifying the `trainer.py` to let it support pytorch's `per-parameter options`, and need modyfing the fp16 relates optimizer, I'll test whether it will be ok =  =

and looking forward the official supports! Because I think it is crucial for fine-tuning when network has pretrained and new-added module.

**Comment by myleott** [comment]
> I want to let part B to get lr1, and let A, C share the same lr2
> following the given method, in my view, I can could make it by
> Is my understanding right?

Yep, that looks right.

> I'm now trying to modify the trainer.py to let it support pytorch's per-parameter options, and need modyfing the fp16 relates optimizer, I'll test whether it will be ok = =
> and looking forward the official supports! Because I think it is crucial for fine-tuning when network has pretrained and new-added module.

Someone actually added this functionality as part of getting T5 implemented: https://github.com/pytorch/fairseq/pull/2172/files. If you search that PR for "param_groups" you can see the needed changes (I haven't tested it yet).

**Comment by memeda** [comment]
OK! I saw the PR, it has changed the `trainer.py` and let the `params` get from the model, I think it is the right way so that we can use our own grouped-params without changing the fairseq main code.
But I haven't seen any changes in `fp16_optimizer.py`. Because my code is a bit older than the latest, so  the grouped-params may have been implemented? If so, this PR may let fairseq support the feature, or we may need continues update it! 
Thanks.
-----
Finally, I think I got the right direction, I'll use my own implementation currently(time is tense) and hope the future fairseq support this properly.

**Comment by Orlllem** [comment]
I was struggling with this problem, and the solution was to use optimizer `composite`. This was used in wav2vec unsupervised 

```
optimizer:
  _name: composite
  groups:
    generator:
      lr: [0.0004]
      lr_float: null
      optimizer:
        _name: adam
        adam_betas: [0.5,0.98]
        adam_eps: 1e-06
        weight_decay: 0
        amsgrad: false
      lr_scheduler:
        _name: fixed
        warmup_updates: 0
    discriminator:
      lr: [ 0.0005 ]
      lr_float: null
      optimizer:
        _name: adam
        adam_betas: [0.5,0.98]
        adam_eps: 1e-06
        weight_decay: 0.0001
        amsgrad: false
      lr_scheduler:
        _name: fixed
        warmup_updates: 0

lr_scheduler: pass_through 
```

You can create your groups 


```
for p in self.generator.parameters():
     p.param_group = "generator"

for p in self.discriminator.parameters():
     p.param_group = "discriminator"
```

best regards

**Comment by b3y0nd** [comment]
> unsupervised

Do you know how to use --composite in the command line?

**Comment by b3y0nd** [comment]
> I was struggling with this problem, and the solution was to use optimizer `composite`. This was used in wav2vec unsupervised
> 
> ```
> optimizer:
>   _name: composite
>   groups:
>     generator:
>       lr: [0.0004]
>       lr_float: null
>       optimizer:
>         _name: adam
>         adam_betas: [0.5,0.98]
>         adam_eps: 1e-06
>         weight_decay: 0
>         amsgrad: false
>       lr_scheduler:
>         _name: fixed
>         warmup_updates: 0
>     discriminator:
>       lr: [ 0.0005 ]
>       lr_float: null
>       optimizer:
>         _name: adam
>         adam_betas: [0.5,0.98]
>         adam_eps: 1e-06
>         weight_decay: 0.0001
>         amsgrad: false
>       lr_scheduler:
>         _name: fixed
>         warmup_updates: 0
> 
> lr_scheduler: pass_through 
> ```
> 
> You can create your groups
> 
> ```
> for p in self.generator.parameters():
>      p.param_group = "generator"
> 
> for p in self.discriminator.parameters():
>      p.param_group = "discriminator"
> ```
> 
> best regards

Do you know how to use --composite in the command line? @Orlllem

**Comment by jiaohuix** [comment]
> Thanks! This method is a bit tricky, and I don't know what it will affect. Let's say: A model is composited by
> 
> ```
> A -> B -> C
> ```
> 
> I want to let part **B** to get lr1, and let **A**, **C** share the same lr2 following the given method, in my view, I can could make it by
> 
> ```python
> def forward(self, xxxx):
>      # let's say B part's lr1 == 0.2, and A, C part lr2 = 0.02, so it has 10 times.
>      x = A()
>      x = GradMultiply.apply(x, 1 / 10) # A part lr. scale to 1/10 based on B => lr = lr2 = 0.02
>      x = B()
>      x = GradMultiply.apply(x, 10) # B part lr. scale to 10 based on C => lr = lr1 = 0.2
>      x = C() # set the global lr = lr2 = 0.02, and C part lr = lr2 = 0.02
> ```
> 
> Is my understanding right?
> 
> I'm now trying to modify the `trainer.py` to let it support pytorch's `per-parameter options`, and need modyfing the fp16 relates optimizer, I'll test whether it will be ok = =
> 
> and looking forward the official supports! Because I think it is crucial for fine-tuning when network has pretrained and new-added module.

Hi, I want to know if I get x2 from bert output through bert_attn and add it to self_attn output x1, when i want to increase the learning rate of bert_attn, should I use GradMultiply on x for x2? what should I do?  thanks!
 ```

def forward(self, x, bert_out):
    x1 = self.self_attn(x)
    if self.bert_gate:
        x2 = self.bert_attn(q=x, k=bert_out, v=bert_out)
        x = x1 + x2
    else:
        x = x1
    return x
 ```

**Comment by sjchasel** [comment]
> > I was struggling with this problem, and the solution was to use optimizer `composite`. This was used in wav2vec unsupervised
> > ```
> > optimizer:
> >   _name: composite
> >   groups:
> >     generator:
> >       lr: [0.0004]
> >       lr_float: null
> >       optimizer:
> >         _name: adam
> >         adam_betas: [0.5,0.98]
> >         adam_eps: 1e-06
> >         weight_decay: 0
> >         amsgrad: false
> >       lr_scheduler:
> >         _name: fixed
> >         warmup_updates: 0
> >     discriminator:
> >       lr: [ 0.0005 ]
> >       lr_float: null
> >       optimizer:
> >         _name: adam
> >         adam_betas: [0.5,0.98]
> >         adam_eps: 1e-06
> >         weight_decay: 0.0001
> >         amsgrad: false
> >       lr_scheduler:
> >         _name: fixed
> >         warmup_updates: 0
> > 
> > lr_scheduler: pass_through 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can create your groups
> > ```
> > for p in self.generator.parameters():
> >      p.param_group = "generator"
> > 
> > for p in self.discriminator.parameters():
> >      p.param_group = "discriminator"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > best regards
> 
> Do you know how to use --composite in the command line? @Orlllem

Same question. Have you solved the problem?

**Comment by sjchasel** [comment]
> OK! I saw the PR, it has changed the `trainer.py` and let the `params` get from the model, I think it is the right way so that we can use our own grouped-params without changing the fairseq main code. But I haven't seen any changes in `fp16_optimizer.py`. Because my code is a bit older than the latest, so the grouped-params may have been implemented? If so, this PR may let fairseq support the feature, or we may need continues update it! Thanks.
> 
> Finally, I think I got the right direction, I'll use my own implementation currently(time is tense) and hope the future fairseq support this properly.

I find that the lr scheduler used in `trainer.py` will unify all the learning rates in param groups into the default value. Have you encountered this problem?
